File: CNS Lab Assignment-4/PlayfairCipher.java
public class PlayfairCipher {

    static char[][] matrix = new char[5][5];

    static void generateMatrix(String key) {
        boolean[] used = new boolean[26];
        key = key.toUpperCase().replace("J", "I");
        int row = 0, col = 0;

        for (char c : key.toCharArray()) {
            if (c >= 'A' && c <= 'Z' && !used[c - 'A']) {
                matrix[row][col++] = c;
                used[c - 'A'] = true;
                if (col == 5) { col = 0; row++; }
            }
        }

        for (char c = 'A'; c <= 'Z'; c++) {
            if (c == 'J') continue;
            if (!used[c - 'A']) {
                matrix[row][col++] = c;
                used[c - 'A'] = true;
                if (col == 5) { col = 0; row++; }
            }
        }
    }

    static int[] findPosition(char c) {
        if (c == 'J') c = 'I';
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (matrix[i][j] == c) return new int[]{i, j};
            }
        }
        return null;
    }

    static String prepareText(String text) {
        text = text.toUpperCase().replace("J", "I").replaceAll("[^A-Z]", "");
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < text.length(); i++) {
            sb.append(text.charAt(i));
            if (i < text.length() - 1 && text.charAt(i) == text.charAt(i + 1)) {
                sb.append('X');
            } else if (i < text.length() - 1) {
                sb.append(text.charAt(++i));
            }
        }

        if (sb.length() % 2 != 0) sb.append('X');
        return sb.toString();
    }

    static String encrypt(String text) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < text.length(); i += 2) {
            int[] a = findPosition(text.charAt(i));
            int[] b = findPosition(text.charAt(i + 1));

            if (a[0] == b[0]) {
                result.append(matrix[a[0]][(a[1] + 1) % 5]).append(matrix[b[0]][(b[1] + 1) % 5]);
            } else if (a[1] == b[1]) {
                result.append(matrix[(a[0] + 1) % 5][a[1]]).append(matrix[(b[0] + 1) % 5][b[1]]);
            } else {
                result.append(matrix[a[0]][b[1]]).append(matrix[b[0]][a[1]]);
            }
        }
        return result.toString();
    }

    static String decrypt(String text) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < text.length(); i += 2) {
            int[] a = findPosition(text.charAt(i));
            int[] b = findPosition(text.charAt(i + 1));

            if (a[0] == b[0]) {
                result.append(matrix[a[0]][(a[1] + 4) % 5]).append(matrix[b[0]][(b[1] + 4) % 5]);
            } else if (a[1] == b[1]) {
                result.append(matrix[(a[0] + 4) % 5][a[1]]).append(matrix[(b[0] + 4) % 5][b[1]]);
            } else {
                result.append(matrix[a[0]][b[1]]).append(matrix[b[0]][a[1]]);
            }
        }
        return result.toString();
    }

    public static void main(String[] args) {
        generateMatrix("SECURITY");
        String input = "HELLOWORLD";
        String prepared = prepareText(input);
        String encrypted = encrypt(prepared);
        String decrypted = decrypt(encrypted);

        System.out.println("Key: SECURITY");
        System.out.println("Plaintext: " + input);
        System.out.println("Prepared: " + prepared);
        System.out.println("Encrypted: " + encrypted);
        System.out.println("Decrypted: " + decrypted);
    }
}

File: CNS Lab Assignment-4/PlayfairCipher.py
matrix = []
def generate_matrix(key):
    global matrix
    key = key.upper().replace("J", "I")
    used = []
    matrix_chars = []
    for ch in key:
        if ch not in used and ch.isalpha():
            used.append(ch)
            matrix_chars.append(ch)
    for ch in "ABCDEFGHIKLMNOPQRSTUVWXYZ":
        if ch not in used:
            used.append(ch)
            matrix_chars.append(ch)
    matrix = [matrix_chars[i:i + 5] for i in range(0, 25, 5)]
def print_matrix():
    print("Playfair Matrix:")
    for row in matrix:
        print(row)
def find_position(ch):
    if ch == 'J':
        ch = 'I'
    for i in range(5):
        for j in range(5):
            if matrix[i][j] == ch:
                return i, j
def prepare_text(text):
    text = text.upper().replace("J", "I")
    prepared = ""
    i = 0
    while i < len(text):
        a = text[i]
        if i + 1 < len(text):
            b = text[i + 1]
            if a == b:
                prepared += a + 'X'
                i += 1
            else:
                prepared += a + b
                i += 2
        else:
            prepared += a + 'X'
            i += 1
    return prepared
def encrypt(text):
    result = ""
    for i in range(0, len(text), 2):
        a = text[i]
        b = text[i + 1]
        row1, col1 = find_position(a)
        row2, col2 = find_position(b)
        if row1 == row2:
            result += matrix[row1][(col1 + 1) % 5]
            result += matrix[row2][(col2 + 1) % 5]
        elif col1 == col2:
            result += matrix[(row1 + 1) % 5][col1]
            result += matrix[(row2 + 1) % 5][col2]
        else:
            result += matrix[row1][col2]
            result += matrix[row2][col1]
    return result
def decrypt(text):
    result = ""
    for i in range(0, len(text), 2):
        a = text[i]
        b = text[i + 1]
        row1, col1 = find_position(a)
        row2, col2 = find_position(b)
        if row1 == row2:
            result += matrix[row1][(col1 - 1) % 5]
            result += matrix[row2][(col2 - 1) % 5]
        elif col1 == col2:
            result += matrix[(row1 - 1) % 5][col1]
            result += matrix[(row2 - 1) % 5][col2]
        else:
            result += matrix[row1][col2]
            result += matrix[row2][col1]
    return result
generate_matrix("MONARCHY")
print_matrix()
plaintext = "INSTRUMENTS"
prepared_text = prepare_text(plaintext)
encrypted_text = encrypt(prepared_text)
print("\nPrepared Text:", prepared_text)
print("Encrypted Text:", encrypted_text)
generate_matrix("SECURITY")
print("\n")
print_matrix()
ciphertext = "FUOQMPXNSPHQ"
decrypted_text = decrypt(ciphertext)
print("\nCipher Text:", ciphertext)
print("Decrypted Text:", decrypted_text)
